compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/BindsMethodValidator.kt

```diff
@@ -54,70 +79,71 @@ internal class BindsMethodValidator : PrivateCodeGenerator() {
   private fun validateBindsFunction(function: MemberFunctionReference.Psi) {
     if (!function.isAbstract()) {
       throw AnvilCompilationExceptionFunctionReference(
-        message = "@Binds methods must be abstract",
+        message = Errors.BINDS_MUST_BE_ABSTRACT,
         functionReference = function,
       )
     }
 
+    val parameters = function.parameters
+    val psiFunction = function.function
     val hasSingleBindingParameter =
-      (function.parameters.size == 1 && !function.function.isExtensionDeclaration()) ||
-        (function.parameters.isEmpty() && function.function.isExtensionDeclaration())
+      (parameters.size == 1 && !psiFunction.isExtensionDeclaration()) ||
+        (parameters.isEmpty() && psiFunction.isExtensionDeclaration())
     if (!hasSingleBindingParameter) {
       throw AnvilCompilationExceptionFunctionReference(
-        message = "@Binds methods must have exactly one parameter, " +
-          "whose type is assignable to the return type",
+        message = Errors.BINDS_MUST_HAVE_SINGLE_PARAMETER,
         functionReference = function,
       )
     }
 
     function.returnTypeOrNull() ?: throw AnvilCompilationExceptionFunctionReference(
-      message = "@Binds methods must return a value (not void)",
+      message = Errors.BINDS_MUST_RETURN_A_VALUE,
       functionReference = function,
     )
+    val returnTypeName = function.returnType().asClassReference().asTypeName()
 
-    if (!function.parameterMatchesReturnType() && !function.receiverMatchesReturnType()) {
-      val returnType = function.returnType().asClassReference().shortName
-      val paramSuperTypes = (function.parameterSuperTypes() ?: function.receiverSuperTypes())!!
-        .map { it.shortName }
-        .toList()
+    val parameterSuperTypes = function.singleParameterTypeOrNull().superTypes()
+    val receiverSuperTypes = function.receiverType().superTypes()
 
-      val superTypesMessage = if (paramSuperTypes.size == 1) {
-        "has no supertypes."
-      } else {
-        "only has the following supertypes: ${paramSuperTypes.drop(1)}"
-      }
-      throw AnvilCompilationExceptionFunctionReference(
-        message = "@Binds methods' parameter type must be assignable to the return type. " +
-          "Expected binding of type $returnType but impl parameter of type " +
-          "${paramSuperTypes.first()} $superTypesMessage",
-        functionReference = function,
-      )
+    when (returnTypeName) {
+      in parameterSuperTypes -> return
+      in receiverSuperTypes -> return
+      else -> Unit
```

Comment:
Nit: since we don't have an else case here it would feel a bit nicer to keep it as an `if (condition || condition) { // fail }` structure, which is also beneficial over the early returns since that means this check isn't required to stay at the bottom of the validations list as we add more or shift things around.

Reply:
Fixed